smart_card_connector_app/src/application_unittest.cc

```diff
@@ -1066,6 +1066,56 @@ TEST_F(SmartCardConnectorApplicationSingleClientTest,
   EXPECT_EQ(return_code, SCARD_E_PROTO_MISMATCH);
 }
 
+// If the client is allowed to use the SCardDisconnect fallback by admin policy,
+// `SCardConnect()` call from JS succeeds to connect via the "T1" protocol
+// even if the previous connection was using the "RAW" protocol.
+TEST_F(SmartCardConnectorApplicationSingleClientTest,
+       SCardConnectProtocolMismatchDisconnectFallback) {
+  // Arrange:
+  TestingSmartCardSimulation::Device device;
+  device.id = 123;
+  device.type = TestingSmartCardSimulation::DeviceType::kGemaltoPcTwinReader;
+  device.card_type = TestingSmartCardSimulation::CardType::kCosmoId70;
+  SetUsbDevices({device});
+  StartApplication();
+  SetUpJsClient();
+  SetUpSCardContext();
+
+  // Simulate an UpdateAdminPolicy message to allowlist the client.
+  std::vector<std::string> scard_disconnect_fallback_client_app_ids;
+  scard_disconnect_fallback_client_app_ids.push_back(kFakeClientNameForLog);
+  SimulateFakeJsMessage(
+      "update_admin_policy",
+      DictValueBuilder()
+          .Add("scard_disconnect_fallback_client_app_ids",
+               std::move(scard_disconnect_fallback_client_app_ids))
+          .Get());
+
+  // Act:
+  // Connect via the "RAW" protocol and disconnect.
+  SCARDHANDLE scard_handle = 0;
+  DWORD active_protocol = 0;
+  EXPECT_EQ(SimulateConnectCallFromJsClient(
+                kFakeHandlerId, scard_context(), kGemaltoPcTwinReaderPcscName0,
+                SCARD_SHARE_SHARED,
+                /*preferred_protocols=*/SCARD_PROTOCOL_RAW, scard_handle,
+                active_protocol),
+            SCARD_S_SUCCESS);
+  EXPECT_NE(scard_handle, 0);
+  EXPECT_EQ(active_protocol, static_cast<DWORD>(SCARD_PROTOCOL_RAW));
+  EXPECT_EQ(SimulateDisconnectCallFromJsClient(kFakeHandlerId, scard_handle,
+                                               SCARD_LEAVE_CARD),
+            SCARD_S_SUCCESS);
+  // Attempt connecting via a different protocol ("ANY" denotes "either T0 or
+  // T1").
+  LONG return_code = SimulateConnectCallFromJsClient(
+      kFakeHandlerId, scard_context(), kGemaltoPcTwinReaderPcscName0,
+      SCARD_SHARE_SHARED, SCARD_PROTOCOL_ANY, scard_handle, active_protocol);
+
+  // Assert:
```

Comment:
Let's also check the protocol is T1, so that the test doesn't pass accidentally if the RAW protocol sticks by mistake.